alpha/declcfg/diff_include.go

```diff
@@ -0,0 +1,78 @@
+package declcfg
+
+import (
+	"github.com/operator-framework/operator-registry/alpha/model"
+)
+
+// makeUpgradeGraph creates a DAG of bundles with map key Bundle.Replaces.
+func makeUpgradeGraph(ch *model.Channel) map[string][]*model.Bundle {
+	graph := map[string][]*model.Bundle{}
+	for _, b := range ch.Bundles {
+		b := b
+		if b.Replaces != "" {
+			graph[b.Replaces] = append(graph[b.Replaces], b)
+		}
+	}
+	return graph
+}
+
+// findIntersectingBundles finds the intersecting bundle of start and end in the
+// replaces upgrade graph graph by traversing down to the lowest graph node,
+// then returns every bundle higher than the intersection. It is possible
+// to find no intersection; this should only happen when start and end
+// are not part of the same upgrade graph.
+// Output bundle order is not guaranteed.
+func findIntersectingBundles(ch *model.Channel, start, end *model.Bundle, graph map[string][]*model.Bundle) ([]*model.Bundle, bool) {
+	// The intersecting set is equal to end if start is end.
+	if start.Name == end.Name {
+		return []*model.Bundle{end}, true
+	}
+
+	// Construct the old replaces chain from start.
+	oldChain := map[string]*model.Bundle{start.Name: nil}
+	for next := start; next != nil && next.Replaces != ""; next = ch.Bundles[next.Replaces] {
+		oldChain[next.Replaces] = next
+	}
+
+	// Trace the new replaces chain from end until the old chain intersects.
+	var intersection string
+	for next := end; next != nil && next.Replaces != ""; next = ch.Bundles[next.Replaces] {
+		if _, inChain := oldChain[next.Replaces]; inChain {
+			intersection = next.Replaces
+			break
+		}
```

Comment:
Should this be checking if `next.Name` is in the chain rather than `next.Replaces`? What if `end` is in the chain but is not equal to `start`?

Follow up (having looked at the tests)... if start and end are accidentally reversed by a caller such that oldChain contains end, what happens?

Reply:
>Should this be checking if next.Name is in the chain rather than next.Replaces? What if end is in the chain but is not equal to start?

Each name will also be in a replaces, so the algorithm is effectively just skipping `next.Name`. 
I'll add a check for `next.Name` as a base case.

>Follow up (having looked at the tests)... if start and end are accidentally reversed by a caller such that oldChain contains end, what happens?

The intersection will always be the same because the algorithm constructs the same DAGs for both, so the set produced by BFS will always be the same too. The case you mention causes an issue when deleting bundles between start and intersection inclusively, since deleting from the true end leaf node (which could be reversed as `start` as input) will remove bundles that should be in the output. However the only `end` ever used in `findIntersectingBundles()` is the channel head, so we can assume that deleting bundles between `start` and the intersection is ok. In general we can assume `start` and `end` are never reversed.

Reply:
> However the only end ever used in findIntersectingBundles() is the channel head, so we can assume that deleting bundles between start and the intersection is ok. In general we can assume start and end are never reversed.

That's fair, I'm just worried about this assumption being implicit and the possibility that we later change the code to violate that assumption. Is there a preventative measure we can take?

For example:

> What if end is in the chain but is not equal to start?

Can we have an explicit check/error (or maybe panic) for this, since we always assume/expect that `end` encompasses `start`?

Reply:
Added a check for end being in start's chain: https://github.com/operator-framework/operator-registry/pull/756/files#diff-f61e326af35ec0122edb28f797b31bdfdd84cddb1975b1d2df38a10ce8292379R41-R42.